# Practical example. Audiobooks

## Problem

You are given data from an Audiobook app. Logically, it relates only to the audio versions of books. Each customer in the database has made a purchase at least once, that's why he/she is in the database. We want to create a machine learning algorithm based on our available data that can predict if a customer will buy again from the Audiobook company.

The main idea is that if a customer has a low probability of coming back, there is no reason to spend any money on advertizing to him/her. If we can focus our efforts ONLY on customers that are likely to convert again, we can make great savings. Moreover, this model can identify the most important metrics for a customer to come back again. Identifying new customers creates value and growth opportunities.

You have a .csv summarizing the data. There are several variables: Customer ID, Book length in mins_avg (average of all purchases), Book length in minutes_sum (sum of all purchases), Price Paid_avg (average of all purchases), Price paid_sum (sum of all purchases), Review (a Boolean variable), Review (out of 10), Total minutes listened, Completion (from 0 to 1), Support requests (number), and Last visited minus purchase date (in days).

So these are the inputs (excluding customer ID, as it is completely arbitrary. It's more like a name, than a number).

The targets are a Boolean variable (so 0, or 1). We are taking a period of 2 years in our inputs, and the next 6 months as targets. So, in fact, we are predicting if: based on the last 2 years of activity and engagement, a customer will convert in the next 6 months. 6 months sounds like a reasonable time. If they don't convert after 6 months, chances are they've gone to a competitor or didn't like the Audiobook way of digesting information. 

The task is simple: create a machine learning algorithm, which is able to predict if a customer will buy again. 

This is a classification problem with two classes: won't buy and will buy, represented by 0s and 1s. 

Good luck!

## Create the machine learning algorithm



### Import the relevant libraries

In [1]:
# we must import the libraries once again since we haven't imported them in this file
import numpy as np
import tensorflow as tf

### Data

In [2]:
# Create a temporary variable npz, where we will store each of the three Audiobooks datasets
npz = np.load('G:/Meu Drive/PersonalProj/GitHub/Python_Statistics/Python_Deep_Learning/Data/Audiobooks_data_train.npz')

# we extract the inputs using the keyword under which we saved them
# to ensure that they are all floats, let's also take care of that
train_inputs = npz['inputs'].astype(np.float32)
# targets must be int because of sparse_categorical_crossentropy (we want to be able to smoothly one-hot encode them)
train_targets = npz['targets'].astype(np.int_)

# we load the validation data in the temporary variable
npz = np.load('G:/Meu Drive/PersonalProj/GitHub/Python_Statistics/Python_Deep_Learning/Data/Audiobooks_data_validation.npz')
# we can load the inputs and the targets in the same line
validation_inputs, validation_targets = npz['inputs'].astype(np.float32), npz['targets'].astype(np.int_)

# we load the test data in the temporary variable
npz = np.load('G:/Meu Drive/PersonalProj/GitHub/Python_Statistics/Python_Deep_Learning/Data/Audiobooks_data_test.npz')
# we create 2 variables that will contain the test inputs and the test targets
test_inputs, test_targets = npz['inputs'].astype(np.float32), npz['targets'].astype(np.int_)

### Model
Outline, optimizers, loss, early stopping and training

In [3]:
# Set the input and output sizes
input_size = 10
output_size = 2
# Use same hidden layer size for both hidden layers. Not a necessity.
hidden_layer_size = 50
    
# define how the model will look like
model = tf.keras.Sequential([
    # tf.keras.layers.Dense is basically implementing: output = activation(dot(input, weight) + bias)
    # it takes several arguments, but the most important ones for us are the hidden_layer_size and the activation function
    tf.keras.layers.Dense(hidden_layer_size, activation='relu'), # 1st hidden layer
    tf.keras.layers.Dense(hidden_layer_size, activation='relu'), # 2nd hidden layer
    
    #tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
    #tf.keras.layers.Dense(hidden_layer_size, activation='relu'),

    # the final layer is no different, we just make sure to activate it with softmax
    tf.keras.layers.Dense(output_size, activation='softmax') # output layer
])


### Choose the optimizer and the loss function

# we define the optimizer we'd like to use, 
# the loss function, 
# and the metrics we are interested in obtaining at each iteration
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

### Training
# That's where we train the model we have built.

# set the batch size
batch_size = 40

# set a maximum number of training epochs
max_epochs = 100

# set an early stopping mechanism
# let's set patience=2, to be a bit tolerant against random validation loss increases
early_stopping = tf.keras.callbacks.EarlyStopping(patience=2)

# fit the model
# note that this time the train, validation and test data are not iterable
model.fit(train_inputs, # train inputs
          train_targets, # train targets
          batch_size=batch_size, # batch size
          epochs=max_epochs, # epochs that we will train for (assuming early stopping doesn't kick in)
          # callbacks are functions called by a task when a task is completed
          # task here is to check if val_loss is increasing
          callbacks=[early_stopping], # early stopping
          validation_data=(validation_inputs, validation_targets), # validation data
          verbose = 2 # making sure we get enough information about the training process
          )  

Epoch 1/100
90/90 - 5s - 56ms/step - accuracy: 0.8318 - loss: 0.4636 - val_accuracy: 0.8926 - val_loss: 0.3010
Epoch 2/100
90/90 - 0s - 4ms/step - accuracy: 0.8896 - loss: 0.3127 - val_accuracy: 0.9060 - val_loss: 0.2647
Epoch 3/100
90/90 - 0s - 4ms/step - accuracy: 0.8924 - loss: 0.2876 - val_accuracy: 0.9150 - val_loss: 0.2415
Epoch 4/100
90/90 - 1s - 6ms/step - accuracy: 0.8975 - loss: 0.2733 - val_accuracy: 0.9239 - val_loss: 0.2400
Epoch 5/100
90/90 - 0s - 3ms/step - accuracy: 0.8994 - loss: 0.2692 - val_accuracy: 0.9172 - val_loss: 0.2210
Epoch 6/100
90/90 - 0s - 4ms/step - accuracy: 0.9022 - loss: 0.2619 - val_accuracy: 0.9262 - val_loss: 0.2287
Epoch 7/100
90/90 - 0s - 4ms/step - accuracy: 0.9003 - loss: 0.2589 - val_accuracy: 0.9262 - val_loss: 0.2056
Epoch 8/100
90/90 - 0s - 3ms/step - accuracy: 0.9030 - loss: 0.2562 - val_accuracy: 0.9284 - val_loss: 0.2144
Epoch 9/100
90/90 - 0s - 4ms/step - accuracy: 0.9039 - loss: 0.2515 - val_accuracy: 0.9262 - val_loss: 0.2092


## Test the model

As we discussed in the lectures, after training on the training data and validating on the validation data, we test the final prediction power of our model by running it on the test dataset that the algorithm has NEVER seen before.

It is very important to realize that fiddling with the hyperparameters overfits the validation dataset. 

The test is the absolute final instance. You should not test before you are completely done with adjusting your model.

If you adjust your model after testing, you will start overfitting the test dataset, which will defeat its purpose.

In [4]:
test_loss, test_accuracy = model.evaluate(test_inputs, test_targets)

14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9119 - loss: 0.2456 


In [5]:
print('\nTest loss: {0:.2f}. Test accuracy: {1:.2f}%'.format(test_loss, test_accuracy*100.))


Test loss: 0.25. Test accuracy: 90.85%


Using the initial model and hyperparameters given in this notebook, the final test accuracy should be roughly around 91%.

Note that each time the code is rerun, we get a different accuracy because each training is different. 

We have intentionally reached a suboptimal solution, so you can have space to build on it!